From 27b6a96febcba5a160234f341e2ebc6ae4a56f42 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 27 Mar 2025 19:39:39 +0100
Subject: [PATCH 1647/1695] plat-stm32mp1: stm32mp1_pwr: fix access to EXTI in
 low-power function

Remove ithe link between the low power function and EXTI driver and need
of the obsolete property "wakeup-parent" for the pwr node.

Since migration of EXTI driver to interrupt framework, we have no reason
to add "wakeup-parent = <&exti>;" in node pwr_regulators: pwr@50001000.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 982cf540cc080a2738bff5d9f5fe61dd55cf1a4d plat-stm32mp1: stm32mp1_pwr: fix access to EXTI in low-power function]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I1b1524d4d9b94ac13086041d3e58102fe1c59106
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/447278
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
---
 core/arch/arm/plat-stm32mp1/pm/power_config.c | 25 +++++++------------
 1 file changed, 9 insertions(+), 16 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/pm/power_config.c b/core/arch/arm/plat-stm32mp1/pm/power_config.c
index cce4d7eae..b84a12c7e 100644
--- a/core/arch/arm/plat-stm32mp1/pm/power_config.c
+++ b/core/arch/arm/plat-stm32mp1/pm/power_config.c
@@ -5,11 +5,11 @@
 
 #include <assert.h>
 #include <config.h>
-#include <drivers/stm32_exti.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <kernel/dt.h>
 #include <kernel/boot.h>
 #include <kernel/panic.h>
+#include <io.h>
 #include <libfdt.h>
 #include <stm32_util.h>
 #include <stm32mp_pm.h>
@@ -101,8 +101,8 @@ static void dump_pm_domain_state(uint8_t domain __unused)
 }
 #else /* CFG_STM32_LOWPOWER_SIP */
 
-static struct stm32_exti_pdata *exti;
 #define EXTI_BANK_NR		3U
+#define EXTI_C1IMR(n)		(0x080U + (n) * 0x10U)
 
 #ifdef CFG_STM32MP13
 /*
@@ -186,13 +186,18 @@ static uint32_t imr_pd_core_ret_mask[EXTI_BANK_NR] = {
 	IMR3_PD_CORE_RET_MASK,
 };
 
+static uint32_t exti_read_imr(unsigned int bank)
+{
+	return io_read32(stm32_exti_base() + EXTI_C1IMR(bank));
+}
+
 static bool get_domain_state_from_exti(uint8_t domain)
 {
 	unsigned int i = 0;
 
 	for (i = 0; i < EXTI_BANK_NR; i++) {
 		uint32_t imr_mask = imr_pd_core_mask[i];
-		uint32_t imr = stm32_exti_read_imr(exti, i);
+		uint32_t imr = exti_read_imr(i);
 
 		if (domain == STM32MP1_PD_CORE_RET)
 			imr_mask |= imr_pd_core_ret_mask[i];
@@ -207,9 +212,6 @@ static bool get_domain_state_from_exti(uint8_t domain)
 /* The function returns FALSE if the domain is in use. */
 static bool get_pm_domain_state(uint8_t domain)
 {
-	if (!exti)
-		return true;
-
 	return get_domain_state_from_exti(domain) == false;
 }
 
@@ -219,7 +221,7 @@ static void dump_pm_domain_state(uint8_t __maybe_unused domain)
 	unsigned int i = 0;
 
 	for (i = 0; i < EXTI_BANK_NR; i++) {
-		uint32_t imr = stm32_exti_read_imr(exti, i);
+		uint32_t imr = exti_read_imr(i);
 		uint32_t imr_mask = imr_pd_core_mask[i];
 		const char *name = "PD_CORE";
 
@@ -376,15 +378,6 @@ static TEE_Result stm32mp1_init_lp_states(void)
 	/* Initialize suspend support to the deepest possible mode */
 	deepest_suspend_mode = STM32_PM_CSTOP_ALLOW_STANDBY_DDR_SR;
 
-#ifndef CFG_STM32_LOWPOWER_SIP
-	res = dt_driver_device_from_node_idx_prop("wakeup-parent",
-						  fdt, pwr_node, 0,
-						  DT_DRIVER_INTERRUPT,
-						  &exti);
-	if (res)
-		return res;
-#endif
-
 #ifdef CFG_STM32MP1_OPTEE_IN_SYSRAM
 	if (!stm32mp_supports_hw_cryp())
 		deepest_suspend_mode = STM32_PM_CSTOP_ALLOW_LPLV_STOP;
-- 
2.47.2

